<!--NewPage--><!-- Created by Masa Maeda and Arturo Bejar on 06.04.97  --><!-- Copyright 1997 Electric Communitites. --><!-- All rights reserved worldwide.  --><!-- Proprietary and confidential. --><HTML><HEAD><pre><a href="ExHndl5.html">Previous</a>  <a href="ExHndl7.html">Next</a>  <a href="ExHndlI.html">Index</a>  <a href="http://www-int.communities.com">EC internal web page</a></pre><TITLE>Exception Handling: Page 6</TITLE></HEAD><BODY><P><HR WIDTH="100%" SIZE=2></P><P>&nbsp;</P><B><P>Burlesque: Recursive Catching</P></B><P>Another alternative is to generate some sort of recursive catching in which, when an exception is thrown, a cascaded throw that recurse back until a successful line of computation is reached. The complexity of this operation makes it hard to test and correct behavior prediction becomes quite a challenge, thus a potential source of deep purple.</P><P>To revert and reach either an active or a successful line of computation, the value of <FONT FACE="Courier">oldID</FONT> is set to 0 and used as a flag to indicate that its is one that failed. Thus, when an exception is thrown, the corresponding <FONT FACE="Courier">ecatch</FONT> either behaves as explained in the 'Andante' section of this document or throws a new exception to continue revertion.</P><P><HR WIDTH="90%" SIZE=3></P><PRE>	eclass Foo {	  int myID;	  method uFoo(Bar bar, Bud bud) {	    myID++;	    int oldID = myID;	    etry {	      bar &lt- uBar(bud);	    } ecatch {	      if(oldID == myID) {	      	oldID = 0;	        exceptionRecoveryCode;	      } else {	        if (oldID == 0) {	          ethrow(new RtEException("reverting computation..."));	        }	      }	    }	  }	}</PRE><P><HR WIDTH="100%" SIZE=2></P><pre><a href="ExHndl5.html">Previous</a>  <a href="ExHndl7.html">Next</a>  <a href="ExHndlI.html">Index</a>  <a href="http://www-int.communities.com">EC internal web page</a></pre></BODY></HTML>